src/lem/circuit.rs

```diff
@@ -0,0 +1,899 @@
+//! ## Constraint system for LEM
+//!
+//! This module implements the generation of bellperson constraints for LEM, such
+//! that it can be used with Nova folding to prove evaluations of Lurk expressions.
+//!
+//! Some LEMs may require expensive gadgets, such as Poseidon hashing. So we use
+//! the concept of "slots" to avoid wasting constraints. To explore this idea,
+//! let's use the following LEM as an example:
+//!
+//! ```text
+//! a b c {
+//!     match_tag c {
+//!         Num => {
+//!             let x: Cons = hash2(a, b);
+//!             return (x, x, x);
+//!         },
+//!         Char => {
+//!             let m: Cons = hash2(b, a);
+//!             let n: Cons = hash2(c, a);
+//!             return (m, m, n);
+//!         }
+//!     }
+//! }
+//! ```
+//!
+//! On a first impression, one might think that we need to perform three hashing
+//! operations in the circuit when in fact we can get away with only two. That
+//! is so because interpretation can only follow one of the paths:
+//!
+//! * If it goes through `Num`, we need to get one hash right
+//! * If it goes through `Char`, we need to get two hashes right
+//!
+//! Either way, that's at most two hashes that we really care about. So we say
+//! that we need to allocate two slots. The first slot is for the the hash of `x`
+//! or `m` and the second slot if for the hash of `n`. Let's see a sketch of part
+//! of the circuit:
+//!
+//! ```text
+//!     ┌─────┐        ┌─────┐
+//! s0i0┤slot0├s0  s1i0┤slot1├s1
+//! s0i1┤hash2│    s1i1┤hash2│
+//!     └─────┘        └─────┘
+//! ...
+//! PNum = c.tag == Num
+//! PChar = c.tag == Char
+//!
+//! PNum → a == s0i0
+//! PNum → b == s0i1
+//! PNum → x.hash == s0
+//!
+//! PChar → b == s0i0
+//! PChar → a == s0i1
+//! PChar → m.hash == s0
+//!
+//! PChar → c == s1i0
+//! PChar → a == s1i1
+//! PChar → n.hash == s1
+//! ```
+//!
+//! `PNum` and `PChar` are boolean premises that indicate whether interpretation
+//! went through `Num` or `Char` respectively. They're used as inputs for gadgets
+//! that implement implications (hence the right arrows above). We will talk
+//! about "concrete" vs "virtual" paths later.
```

Comment:
Maybe 'elsewhere' would be better than 'later'? I'm not sure how much sequentiality should be assumed in how readers will consume the documentation; and this theme is not finalized in the module doc.

Reply:
Will change